File: frontend/lib/capabilities.ts
export type Capability =
  | 'view_all_operations'
  | 'manage_department'
  | 'make_decisions'
  | 'manage_accounts'
  | 'manage_system'
  | 'view_system_health'
  | 'manage_approvals'
  | 'view_sensitive_user_metadata'
  | 'view_integration_summary';

export function normalizeRole(role: unknown) {
  return String(role || '').trim().toLowerCase();
}

function fromNames(names: unknown[]): Record<string, boolean> {
  return Object.fromEntries(
    names.filter((name): name is string => typeof name === 'string').map(name => [name, true]),
  );
}

function trueBooleanEntries(value: Record<string, unknown>): Record<string, boolean> {
  const out: Record<string, boolean> = {};
  Object.entries(value).forEach(([name, granted]) => {
    if (granted === true) out[name] = true;
  });
  return out;
}

function capabilityMap(source: unknown): Record<string, boolean> {
  if (!source) return {};
  if (Array.isArray(source)) return fromNames(source);
  if (typeof source !== 'object') return {};

  const record = source as Record<string, unknown>;
  const nested = record.capabilities;

  if (Array.isArray(nested)) return fromNames(nested);
  if (nested && typeof nested === 'object') {
    return trueBooleanEntries(nested as Record<string, unknown>);
  }

  return trueBooleanEntries(record);
}

export function hasCapability(source: unknown, capability: Capability) {
  return capabilityMap(source)[capability] === true;
}

File: backend/app/capabilities.py
from __future__ import annotations

from typing import Final

CAPABILITIES: Final[tuple[str, ...]] = (
    "view_all_operations",
    "manage_department",
    "make_decisions",
    "manage_accounts",
    "manage_system",
    "view_system_health",
    "manage_approvals",
    "view_sensitive_user_metadata",
    "view_integration_summary",
)

ROLE_CAPABILITIES: Final[dict[str, frozenset[str]]] = {
    "owner": frozenset(CAPABILITIES),
    "admin": frozenset({
        "view_all_operations",
        "manage_department",
        "make_decisions",
        "manage_accounts",
        "manage_system",
        "view_system_health",
        "manage_approvals",
        "view_sensitive_user_metadata",
        "view_integration_summary",
    }),
    "manager": frozenset({
        "view_all_operations",
        "manage_department",
        "make_decisions",
        "view_system_health",
        "manage_approvals",
        "view_integration_summary",
    }),
    "lead": frozenset({"manage_department"}),
    "supervisor": frozenset({"manage_department"}),
    "staff": frozenset(),
}


def normalize_role(role: object) -> str:
    return str(role or "").strip().lower()


def capabilities_for_role(role: object) -> frozenset[str]:
    return ROLE_CAPABILITIES.get(normalize_role(role), frozenset())


def has_capability(role: object, capability: str) -> bool:
    return capability in capabilities_for_role(role)


def capability_payload(role: object) -> dict[str, bool]:
    granted = capabilities_for_role(role)
    return {capability: capability in granted for capability in CAPABILITIES}
